src/H5RS.c

```diff
@@ -85,26 +106,146 @@ H5FL_BLK_DEFINE(str_buf);
  EXAMPLES
  REVISION LOG
 --------------------------------------------------------------------------*/
-static char *
-H5RS__xstrdup(const char *s)
+static herr_t
+H5RS__xstrdup(H5RS_str_t *rs, const char *s)
 {
-    char *ret_value; /* Return value */
+    herr_t ret_value = SUCCEED; /* Return value */
 
-    FUNC_ENTER_STATIC_NOERR
+    FUNC_ENTER_STATIC
 
-    if (s) {
-        size_t len = HDstrlen(s) + 1;
+    /* Sanity check */
+    HDassert(rs);
 
-        ret_value = (char *)H5FL_BLK_MALLOC(str_buf, len);
-        HDassert(ret_value);
-        HDstrncpy(ret_value, s, len);
+    if (s) {
+        size_t len = HDstrlen(s);
+
+        /* Determine size of buffer to allocate */
+        rs->max = H5RS_ALLOC_SIZE;
+        while ((len + 1) > rs->max)
+            rs->max *= 2;
+
+        /* Allocate the underlying string */
+        if (NULL == (rs->s = (char *)H5FL_BLK_MALLOC(str_buf, rs->max)))
+            HGOTO_ERROR(H5E_RS, H5E_CANTALLOC, FAIL, "memory allocation failed")
```

Comment:
I'm not clear when is H5E_RESOURCE and when is not?  H5E_RS here.

Reply:
> I'm not clear when is H5E_RESOURCE and when is not? H5E_RS here.

I suggest using H5E_RESOURCE in the actual resource allocation routines (like H5MM) and the current package (H5E_RS, here) in the package that calls the resource allocation routine.

Reply:
Ah, OK, thanks!

Reply:
I think H5E_RESOURCE should go away. Packages should just use their own major error code.

I hope to eventually refactor the error macros so that the package bit happens automatically and then that would always be the case.

Reply:
> I think H5E_RESOURCE should go away. Packages should just use their own major error code.
> 
> I hope to eventually refactor the error macros so that the package bit happens automatically and then that would always be the case.

Agree - I think all the "generic" errors should go away and we should shift to package-based major error codes.   (And I would to also have them happen automatically, which the module headers could help with)